### PR TITLE
fix UTF-16 to UTF-8 index translation

### PR DESCRIPTION
**Pull request by omg**
This addresses incorrect index translation affecting strings containing characters from Unicode Supplemental Planes (U+10000 - U+10FFFF). These characters use 2 code units in UTF-16 and counted as 2 "characters" in V8, such that `'🤡'.length === 2`. On the other hand in UTF-8 they are encoded using just one 4-byte unit. This caused the code to skip one extra character on each occurrence. For example  `'🤡abc'[2] === 'a'` but when translated the index would point to `b`

i don't say that `getUtf8Length` and `getUtf16Length` functions work incorrectly by themselves, i say the specific code targeted by this commit (doing index translation) uses their results without regard for underlying semantics. Built-in regex has no such problem. In your examples this code has no chance to fail because `lastIndex == 0`, try matching with g flag:

```
> re2 = new RE2(/\d/g)

> re2.exec('🤡123')
[ '1', index: 2, input: '🤡123', groups: undefined ]
> re2.exec('🤡123')
[ '3', index: 3, input: '🤡123', groups: undefined ]
```
where's 2?

built-in regex works fine:
```
> let re = /\d/g

> re.exec('🤡123')
[ '1', index: 2, input: '🤡123', groups: undefined ]
> re.exec('🤡123')
[ '2', index: 3, input: '🤡123', groups: undefined ]
> re.exec('🤡123')
[ '3', index: 4, input: '🤡123', groups: undefined ]
```

In the supplied test case second exec would incorrectly match 'own' before fix, instead of 'clown'

### CODE REVIEW

**Comment by uhop** [comment]
Now I understand the problem. Thank you!

**Comment by uhop** [comment]
I decided to implement it and got the following problem:

```text
$ node
Welcome to Node.js v17.2.0.
Type ".help" for more information.
> 'яя'.length
2
> 'яя'.substr(0, 1)
'я'
> 'яяяя'.substr(0, 2)
'яя'
> '台湾'.length
2
> '台湾'.substr(0, 1)
'台'
> '台湾'.substr(0, 2)
'台湾'
> '🤡🤡'.length
4
> '🤡🤡'.substr(0, 1)
'�'
> '🤡🤡'.substr(0, 2)
'🤡'
> '🤡🤡'.substr(0, 3)
'🤡�'
> '🤡🤡'.substr(0, 4)
'🤡🤡'
```

Clowns are pairs of UTF-16 symbols. Probably we should calculate them as such.

Using `RE2.getUtf16Length()`:

```text
> buf = Buffer.from('яя')
<Buffer d1 8f d1 8f>
> RE2.getUtf16Length(buf)
2
> buf = Buffer.from('台湾')
<Buffer e5 8f b0 e6 b9 be>
> RE2.getUtf16Length(buf)
2
> buf = Buffer.from('🤡🤡')
<Buffer f0 9f a4 a1 f0 9f a4 a1>
> RE2.getUtf16Length(buf)
4
```

It appears to be consistent with `length`.

I need to check what `RegExp` is doing with characters like that.

**Comment by uhop** [comment]
Yes, it is simple, and your code is correct. I never doubted it but wanted to understand it myself.

* UTF-8 table:
  | First code point | Last code point | Length in bytes |
  |----------------:|----------------:|----------------:|
  | U+0000 | U+007F | 1 |  
  | U+0080 | U+07FF | 2 |  
  | U+0800 | U+FFFF | 3 |  
  | U+10000 | U+10FFFF | 4 |
* UTF-16 table:
  | First code point | Last code point | Length in bytes | Length in chars |
  |----------------:|----------------:|----------------:|-----------------:|
  | U+0000 | U+D7FF | 2 |  1 |
  | U+E000 | U+FFFF | 2 |  1 |
  | U+010000 | U+10FFFF | 4 |  2 |

Basically, it follows from the tables that if a character has 4 bytes in UTF-8, it will take 2 characters (4 bytes) in UTF-16.
